src/dune_trace/buffer.ml

```diff
@@ -0,0 +1,44 @@
+open Stdune
+
+type t =
+  { mutable buf : Bigstringaf.t
+  ; mutable pos : int
+  }
+
+let create size =
+  let buf = Bigstringaf.create size in
+  { buf; pos = 0 }
+;;
+
+let max_size t = Bigstringaf.length t.buf
+let available t = Bigstringaf.length t.buf - t.pos
+let clear t = t.pos <- 0
+let to_string t = Bigstringaf.substring t.buf ~off:0 ~len:t.pos
+let buf t = t.buf
+let pos t = t.pos
+
+let add_char t c =
+  Bigstringaf.set t.buf t.pos c;
+  t.pos <- t.pos + 1
+;;
+
+let add_string t str =
+  let len = String.length str in
+  Bigstringaf.blit_from_string str t.buf ~dst_off:t.pos ~src_off:0 ~len;
+  t.pos <- t.pos + len
+;;
+
+let drop t n =
+  assert (t.pos >= n);
+  let buf = Bigstringaf.create (Bigstringaf.length t.buf) in
+  Bigstringaf.blit t.buf ~src_off:n buf ~dst_off:0 ~len:(t.pos - n);
+  t.buf <- buf;
+  t.pos <- t.pos - n
+;;
```

Comment:
I think we can do better here, but it doesn't have to be in this PR. Looking at the documentation of `Bigstringaf` I see we have `unsafe_blit` based on `memmove` rather than `memcpy`. That might save some work for the GC. Maybe make a CR about that?